### PR TITLE
 Replace filter

### PR DESCRIPTION
**Pull request by MarkGibbons**
### Description

Add the replace filter

### Issues Resolved

Adds new function.

### Contribution Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable

---
This change is [Reviewable](https://reviewable.io/reviews/sous-chefs/line/136)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,8 +1,12 @@
 # line Cookbook CHANGELOG
 
-# Unreleased
+# Unreleased - for target version 2.3.0 in mid Janary 2018
```
Comment (tas50) [inline CHANGELOG.md]:
```suggestion
# Unreleased - for target version 2.3.0 in mid January 2019
```

Comment (MarkGibbons) [reply CHANGELOG.md]: Fixed

File: spec/unit/library/filter/after_spec.rb [hunk 2]
```diff
@@ -97,4 +98,16 @@
   it 'should insert a string' do
     expect(@filt.after(@solo_middle, [@pattern_c1, 'string1', :last])).to eq(%w(linef c1 string1 lineg))
   end
+
+  it 'should insert a line that matches the pattern by default, nil implies safe' do
```
Comment (JohnRoesler) [inline spec/unit/library/filter/after_spec.rb]: should not?

File: spec/unit/library/filter/replace_spec.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+  it 'should replace the middle line' do
+    expect(@filt.replace(@solo_middle, [@pattern_c1, @ia])).to eq(%w(linef line1 line2 line3 lineg))
+  end
+
+  it 'should replace the end line' do
+    expect(@filt.replace(@solo_end, [@pattern_c1, @ia])).to eq(%w(linef lineg lineh line1 line2 line3))
+  end
+
+  it 'should raise error if the pattern matchs the replacement lines' do
```
Comment (JohnRoesler) [inline spec/unit/library/filter/replace_spec.rb]:
```suggestion
  it 'should raise error if the pattern matches the replacement lines' do
```

File: spec/unit/library/filter/replace_spec.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    expect { @filt.replace(@current, [@pattern_c1, ['c1 match', 'c2']]) }.to raise_error(ArgumentError)
+  end
+
+  it 'should not raise error if the pattern matchs the replacement lines, force the change' do
```
Comment (JohnRoesler) [inline spec/unit/library/filter/replace_spec.rb]:
```suggestion
  it 'should not raise error if the pattern matches the replacement lines, force the change' do
```

**Review by JohnRoesler — CHANGES_REQUESTED** [review]
Some conflicts to clean up also

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big Pull Request.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by JohnRoesler — APPROVED** [review]

**Pull request merged**